lib/Dialect/DC/DCOps.cpp

```diff
@@ -0,0 +1,195 @@
+//===- DCOps.cpp ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/DC/DCOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/FunctionImplementation.h"
+#include "mlir/IR/OpImplementation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+
+using namespace circt;
+using namespace dc;
+using namespace mlir;
+
+bool circt::dc::isI1ValueType(Type t) {
+  auto vt = t.dyn_cast<ValueType>();
+  if (!vt || vt.getInnerTypes().size() != 1)
+    return false;
+
+  auto innerWidth = vt.getInnerTypes()[0].getIntOrFloatBitWidth();
+  return innerWidth == 1;
+}
+
+namespace circt {
+namespace dc {
+#include "circt/Dialect/DC/DCCanonicalization.h.inc"
+
+// =============================================================================
+// JoinOp
+// =============================================================================
+
+void JoinOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                         MLIRContext *context) {}
+
+// =============================================================================
+// ForkOp
+// =============================================================================
+
+template <typename TInt>
+static ParseResult parseIntInSquareBrackets(OpAsmParser &parser, TInt &v) {
+  if (parser.parseLSquare() || parser.parseInteger(v) || parser.parseRSquare())
+    return failure();
+  return success();
+}
+
+ParseResult ForkOp::parse(OpAsmParser &parser, OperationState &result) {
+  OpAsmParser::UnresolvedOperand operand;
+  size_t size = 0;
+  if (parseIntInSquareBrackets(parser, size))
+    return failure();
+
+  if (size == 0)
+    return parser.emitError(parser.getNameLoc(),
+                            "fork size must be greater than 0");
+
+  if (parser.parseOperand(operand) ||
+      parser.parseOptionalAttrDict(result.attributes))
+    return failure();
+
+  auto tt = dc::TokenType::get(parser.getContext());
+  llvm::SmallVector<Type> operandTypes{tt};
+  SmallVector<Type, 1> resultTypes{size, tt};
```

Comment:
Nit: is 1 really a good assumption? A fork with one output would be kind of silly. I think 8 or just leaving it off would be more appropriate.